docs/spec/overload.rst

```diff
@@ -99,8 +109,396 @@ variable is not sufficient.
+      # without success.
+      example2(val, val, val)  # Error in step 3
+
+
+Example 3::
+
+  @overload
+  def example3(x: int) -> int: ...
+  @overload
+  def example3(x: int, y: int) -> tuple[int, int]: ...
+  @overload
+  def example3(*args: int) -> tuple[int, ...]: ...
+
+  def test(val: list[int]):
+      # Step 1 eliminates second overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the first
+      # overload.
+      r1 = example3(1)
+      reveal_type(r1)  # Should reveal int
+
+      # Step 1 eliminates first overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the second
+      # overload.
+      r2 = example3(1, 2)
+      reveal_type(r2)  # Should reveal tuple[int, int]
+
+      # Step 1 doesn't eliminate any overloads. Step 4
+      # picks the third overload.
+      r3 = example3(*val)
+      reveal_type(r3)  # Should reveal tuple[int, ...]
+
+
+Example 4::
+
+  @overload
+  def example4(x: list[int], y: int) -> int: ...
+  @overload
+  def example4(x: list[str], y: str) -> int: ...
+  @overload
+  def example4(x: int, y: int) -> list[int]: ...
+
+  def test(v1: list[Any], v2: Any):
+      # Step 2 eliminates the third overload. Step 5
+      # determines that first and second overloads
+      # both apply and are ambiguous due to Any, but
+      # return types are consistent.
+      r1 = example4(v1, v2)
+      reveal_type(r1)  # Reveals int
+
+      # Step 2 eliminates the second overload. Step 5
+      # determines that first and third overloads
+      # both apply and are ambiguous due to Any, and
+      # the return types are inconsistent.
+      r2 = example4(v2, 1)
+      reveal_type(r2)  # Should reveal Any
+
+
+.. _argument-type-expansion:
+
+Argument type expansion
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When performing argument type expansion, the following types should be
+expanded:
+
+1. Unions: Each subtype of the union should be considered as a separate
+argument type. For example, the type ``int | str`` should be expanded
+into ``int`` and ``str``.
+
+2. ``bool`` should be expanded into ``Literal[True]`` and ``Literal[False]``.
+
+3. ``Enum`` types (other than those that derive from ``enum.Flag``) should
+be expanded into their literal members.
+
+4. ``type[A | B]`` should be expanded into ``type[A]`` and ``type[B]``.
+
+5. Tuples of known length that contain expandable types should be expanded
+into all possible combinations of their subtypes. For example, the type
+``tuple[int | str, bool]`` should be expanded into ``(int, Literal[True])``,
+``(int, Literal[False])``, ``(str, Literal[True])``, and
+``(str, Literal[False])``.
+
+
+[Eric's note for reviewers: I'm not 100% convinced we should
+support argument expansion in all of these cases. Tuple expansion,
+in particular, can be very costly and can quickly blow up in complexity.
+Currently, pyright and mypy support only the case 1 in the list above,
+but I have had requests to support 2 and 3.]
```

Comment:
The downside to supporting all of this is that overload matching is already insanely complicated in Python — enough so that it takes pages of text to attempt to describe the algorithm. Contrast that to overload matching in most other languages (like TypeScript) where it simply evaluates overloads one at a time and picks the first match. I'm really looking for ways we can make this simpler. If I could go back in time, I'd try to convince the authors of mypy to not add all of this complexity. But now we're kind of stuck with much of it. I'm really hesitant to pile on more complexity.

Reply:
Having personally implemented the overload logic myself recently I also feel like adding more complexity is a bad idea. Mypy also does union math in some places that could lead to pretty slow exponential calls if bools were always unpacked.

Before implementing overloads I thought that this would be pretty easy, but after doing it I feel like it's probably the most complicated part of Python's typing system.

Reply:
I agree that overloads are complicated. But I think that once we have the current complex algorithm, with union expansion already in it, clarifying more forms of types that are equivalent to unions is not a significant addition to the overall complexity of the algorithm. Although it can add complexity to a concrete implementation, in principle I see it as rather a simplification, because equivalent types should not be specified in ways that make them behave as not-equivalent.

If there is major concern about supporting these in existing type checkers, I would at least like for the specification to be clear that all of these types are equivalent to unions, and thus it is permissible for a type checker to expand them as unions. There should not be any example in the conformance suite or any language in the spec suggesting that such an expansion would be wrong.

Reply:
Yes, I think it's an important principle that if two types are equivalent, it should be possible to use them interchangeably. I don't know if it's always feasible for type checkers to follow this principle, but it's still something we should aim towards.

Reply:
Yes, these are good arguments. I've updated the document to spell out the general principle. I still give this list as specific examples, but I make it clear that this list might not be exhaustive. I think it is exhaustive today, but as we extend the type system — e.g. `TypeForm`, there may be additional examples. The same principle should cover all such cases in the future.